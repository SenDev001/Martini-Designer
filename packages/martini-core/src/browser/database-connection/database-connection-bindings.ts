import { CommandContribution, MenuContribution } from "@theia/core";
import { OpenHandler, WebSocketConnectionProvider, WidgetFactory } from "@theia/core/lib/browser";
import { interfaces } from "inversify";
import {
    MartiniDatabaseConnectionManager,
    martiniDatabaseConnectionManagerPath
} from "../../common/database-connection/martini-database-connection-manager";
import {
    MartiniMongoDbManager,
    martiniMongoDbManagerPath
} from "../../common/database-connection/mongodb/martini-mongodb-manager";
import {
    MartiniDatabaseSchemaProvider,
    martiniDatabaseSchemaProviderPath
} from "../../common/database-connection/schema/martini-database-schema-provider";
import {
    NavigatorContentProviderContribution,
    NavigatorLabelProviderContribution,
    NavigatorOpenHandler
} from "../navigator/martini-navigator-view-widget";
import { WizardContribution } from "../wizard/wizard-contribution";
import {
    DatabaseConnectionCommandContribution,
    DatabaseConnectionMenuContribution
} from "./database-connection-contribution";
import {
    DatabaseConnectionEventDispatcher,
    DatabaseConnectionEventDispatcherClient
} from "./database-connection-event-dispatcher";
import {
    DatabaseConnectionNavigatorContentProvider,
    DatabaseConnectionNavigatorLabelProvider
} from "./database-connection-navigator-contribution";
import {
    DatabaseConnectionEditor,
    DatabaseConnectionEditorNavigatorOpenHandler,
    DatabaseConnectionEditorOpenHandler,
    DatabaseConnectionEditorOptions
} from "./editor/database-connection-editor";
import {
    DatabaseConnectionWizard,
    DatabaseConnectionWizardContribution,
    DatabaseConnectionWizardPage
} from "./wizard/database-connection-wizard";

export const bindDatabaseConnectionBindings = (bind: interfaces.Bind) => {
    bind(DatabaseConnectionEventDispatcherClient)
        .toSelf()
        .inSingletonScope();
    bind(DatabaseConnectionEventDispatcher).toService(DatabaseConnectionEventDispatcherClient);
    bind(MartiniDatabaseConnectionManager)
        .toDynamicValue(ctx => {
            const connection = ctx.container.get(WebSocketConnectionProvider);
            return connection.createProxy(
                martiniDatabaseConnectionManagerPath,
                ctx.container.get(DatabaseConnectionEventDispatcherClient)
            );
        })
        .inSingletonScope();
    bind(MartiniDatabaseSchemaProvider)
        .toDynamicValue(ctx => {
            const connection = ctx.container.get(WebSocketConnectionProvider);
            return connection.createProxy(martiniDatabaseSchemaProviderPath);
        })
        .inSingletonScope();
    bind(MartiniMongoDbManager).toDynamicValue(ctx => {
        const connection = ctx.container.get(WebSocketConnectionProvider);
        return connection.createProxy(martiniMongoDbManagerPath);
    });
    bind(NavigatorContentProviderContribution).to(DatabaseConnectionNavigatorContentProvider);
    bind(NavigatorLabelProviderContribution).to(DatabaseConnectionNavigatorLabelProvider);
    bind(CommandContribution)
        .to(DatabaseConnectionCommandContribution)
        .inSingletonScope();
    bind(MenuContribution)
        .to(DatabaseConnectionMenuContribution)
        .inSingletonScope();

    bind(WizardContribution)
        .to(DatabaseConnectionWizardContribution)
        .inSingletonScope();
    bind(DatabaseConnectionWizard).toSelf();
    bind(DatabaseConnectionWizardPage).toSelf();
    bind("Factory<DatabaseConnectionWizard>").toAutoFactory(DatabaseConnectionWizard);

    bind(WidgetFactory)
        .toDynamicValue(({ container }) => ({
            id: DatabaseConnectionEditor.FACTORY_ID,
            createWidget: (options: DatabaseConnectionEditorOptions) => {
                const child = container.createChild();
                child.bind(DatabaseConnectionEditorOptions).toConstantValue(options);
                child.bind(DatabaseConnectionEditor).toSelf();
                return child.get(DatabaseConnectionEditor);
            }
        }))
        .inSingletonScope();

    bind(OpenHandler)
        .to(DatabaseConnectionEditorOpenHandler)
        .inSingletonScope();
    bind(NavigatorOpenHandler)
        .to(DatabaseConnectionEditorNavigatorOpenHandler)
        .inSingletonScope();
};
