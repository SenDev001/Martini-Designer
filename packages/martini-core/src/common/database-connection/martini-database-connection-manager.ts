import { JsonRpcServer } from "@theia/core";
import { MartiniEvent } from "../event/martini-event";
import { startCase } from "lodash";

export enum DatabaseType {
    CASSANDRA = "cassandra",
    JDBC = "jdbc",
    MONGODB = "mongodb",
    REDIS = "redis"
}

export interface DatabaseConnection {
    name: string;
    autoStart: boolean;
    status: "STARTED" | "STOPPED";
    type: DatabaseType;
}

export namespace DatabaseConnection {
    export function is(object: any): object is DatabaseConnection {
        return (
            !!object &&
            typeof object === "object" &&
            "name" in object &&
            "autoStart" in object &&
            "status" in object &&
            "type" in object
        );
    }
    export function getType(type: string): DatabaseType | undefined {
        const typeKey: string | undefined = Object.keys(DatabaseType).find(
            (key: string) => DatabaseType[key as keyof typeof DatabaseType] === type
        );
        return typeKey ? DatabaseType[typeKey as keyof typeof DatabaseType] : undefined;
    }
}

export const martiniDatabaseConnectionManagerPath = "/services/martini/database/connections";

export const MartiniDatabaseConnectionManager = Symbol("MartiniDatabaseConnectionManager");

export interface MartiniDatabaseConnectionManager extends JsonRpcServer<MartiniDatabaseConnectionManagerClient> {
    getAll(): Promise<DatabaseConnection[]>;

    get(name: string): Promise<DatabaseConnection>;

    start(name: string): Promise<void>;

    stop(name: string): Promise<void>;

    delete(name: string): Promise<void>;

    save(connection: DatabaseConnection): Promise<void>;

    test(connection: DatabaseConnection): Promise<void>;

    getDrivers(): Promise<Driver[]>;
}

export interface DatabaseConnectionEvent extends MartiniEvent {
    event: "SAVED" | "DELETED" | "STARTED" | "STOPPED" | "ENABLED" | "DISABLED";
    name: string;
}

export namespace DatabaseConnectionEvent {
    export function is(event: MartiniEvent): event is DatabaseConnectionEvent {
        return event.type === "DATABASE_CONNECTION";
    }
}

export interface MartiniDatabaseConnectionManagerClient {
    onEvent(event: DatabaseConnectionEvent): void;
}

export const getTypeLabel = (type: DatabaseType) => {
    if (type === DatabaseType.JDBC) return "JDBC";
    return startCase(type.toLowerCase());
};

export enum JdbcIsolationLevel {
    READ_COMMITTED = "READ_COMMITTED",
    READ_UNCOMMITTED = "READ_UNCOMMITTED",
    REPEATABLE_READ = "REPEATABLE_READ",
    SERIALIZABLE = "SERIALIZABLE"
}

export interface Driver {
    templateUrl: string;
    name: string;
    className: string;
    xaClassName: string;
}

export interface JdbcDatabaseConnection extends DatabaseConnection {
    type: DatabaseType.JDBC;
    url: string;
    xa: boolean;
    driverClassName: string;
    username?: string;
    password: string;
    acquireIncrement: number;
    acquisitionInterval: number;
    acquisitionTimeout: number;
    allowLocalTransactions: boolean;
    applyTransactionTimeout: boolean;
    automaticEnlistingEnabled: boolean;
    deferConnectionRelease: boolean;
    enableJdbc4ConnectionTest: boolean;
    ignoreRecoveryFailures: boolean;
    isolationLevel: JdbcIsolationLevel;
    maxIdleTime: number;
    minPoolSize: number;
    maxPoolSize: number;
    preparedStatementCacheSize: number;
    shareTransactionConnections: boolean;
    testQuery?: string;
    twoPcOrderingPosition: number;
    uniqueName?: string;
    useTmJoin: boolean;
    loginTimeout: number;
}

export enum CassandraCompression {
    NONE = "NONE",
    SNAPPY = "SNAPPY",
    LZ4 = "LZ4"
}

export enum CassandraConsistencyLevel {
    ANY = "ANY",
    ONE = "ONE",
    TWO = "TWO",
    THREE = "THREE",
    QUORUM = "QUORUM",
    ALL = "ALL",
    LOCAL_QUORUM = "LOCAL_QUORUM",
    EACH_QUORUM = "EACH_QUORUM",
    SERIAL = "SERIAL",
    LOCAL_SERIAL = "LOCAL_SERIAL",
    LOCAL_ONE = "LOCAL_ONE"
}

export interface CassandraDatabaseConnection extends DatabaseConnection {
    type: DatabaseType.CASSANDRA;
    contactPoints: string[];
    port: number;
    maxSchemaAgreementWaitSeconds: number;
    credentials: {
        username?: string;
        password: string;
    };
    compression: CassandraCompression;
    metrics: boolean;
    ssl: boolean;
    poolingOptions: {
        coreConnectionsPerHostLocal: number;
        coreConnectionsPerHostRemote: number;
        maxConnectionsPerHostLocal: number;
        maxConnectionsPerHostRemote: number;
        newConnectionThresholdLocal: number;
        newConnectionThresholdRemote: number;
        maxRequestsPerConnectionLocal: number;
        maxRequestsPerConnectionRemote: number;
        idleTimeoutSeconds: number;
        poolTimeoutMillis: number;
        maxQueueSize: number;
        heartbeatIntervalSeconds: number;
    };
    socketOptions: {
        connectTimeoutMillis: number;
        readTimeoutMillis: number;
        keepAlive: boolean;
        reuseAddress: boolean;
        soLinger: number;
        tcpNoDelay: boolean;
        receiveBufferSize: number;
        sendBufferSize: number;
    };
    queryOptions: {
        consistencyLevel: CassandraConsistencyLevel;
        serialConsistencyLevel: CassandraConsistencyLevel;
        fetchSize: number;
        defaultIdempotence: boolean;
        prepareOnAllHosts: boolean;
        reprepareOnUp: boolean;
        refreshSchemaIntervalMillis: number;
        maxPendingRefreshSchemaRequests: number;
        refreshNodeListIntervalMillis: number;
        maxPendingRefreshNodeListRequests: number;
        refreshNodeIntervalMillis: number;
        maxPendingRefreshNodeRequests: number;
    };
}

export enum MongoDbAuthenticationMechanism {
    AUTO = "AUTO",
    PLAIN = "PLAIN",
    SCRAM_SHA_1 = "SCRAM_SHA_1",
    SCRAM_SHA_256 = "SCRAM_SHA_256"
}

export enum MongoDbClusterType {
    STANDALONE = "STANDALONE",
    REPLICA_SET = "REPLICA_SET",
    SHARDED = "SHARDED",
    UNKNOWN = "UNKNOWN"
}

export interface MongoDbDatabaseConnection extends DatabaseConnection {
    type: DatabaseType.MONGODB;
    connectionString: string;
    username?: string;
    password: string;
    authenticationDatabase?: string;
    authenticationMechanism: MongoDbAuthenticationMechanism;
    clusterSettings: {
        localThreshold: number;
        maxWaitQueueSize: number;
        serverSelectionTimeout: number;
        type: MongoDbClusterType;
        hosts: string[];
    };
    connectionPoolSettings: {
        maintenanceFrequency: number;
        maintenanceInitialDelay: number;
        minSize: number;
        maxConnectionIdleTime: number;
        maxConnectionLifeTime: number;
        maxWaitTime: number;
        maxWaitQueueSize: number;
        maxSize: number;
    };
    serverSettings: {
        heartbeatFrequency: number;
        minHeartbeatFrequency: number;
    };
    socketSettings: {
        connectTimeout: number;
        readTimeout: number;
        receiveBufferSize: number;
        sendBufferSize: number;
    };
}

export enum RedisDisconnectedBehavior {
    DEFAULT = "DEFAULT",
    ACCEPT_COMMANDS = "ACCEPT_COMMANDS",
    REJECT_COMMANDS = "REJECT_COMMANDS"
}
export enum RedisSslProvider {
    JDK = "JDK",
    OPENSSL = "OPENSSL"
}
export interface RedisDatabaseConnection extends DatabaseConnection {
    type: DatabaseType.REDIS;
    uri: string;
    password: string;
    clientOptions: {
        pingBeforeActivateConnection: boolean;
        autoReconnect: boolean;
        cancelCommandsOnReconnectFailure: boolean;
        publishOnScheduler: boolean;
        suspendReconnectOnProtocolFailure: boolean;
        requestQueueSize: number;
        disconnectedBehavior: RedisDisconnectedBehavior;
        socketOptions: {
            connectTimeoutMillis: number;
            keepAlive: boolean;
            tcpNoDelay: boolean;
        };
        sslOptions: {
            sslProvider: RedisSslProvider;
            keystoreFile?: string;
            keystorePassword?: string;
            truststoreFile?: string;
            truststorePassword?: string;
            startTls: boolean;
        };
        connectionOptions: {
            autoFlushCommands: boolean;
            connectionTimeoutMillis: number;
        };
        connectionPoolOptions: {
            maxTotal: number;
            maxIdle: number;
            minIdle: number;
            lifo: boolean;
            fairness: boolean;
            maxWaitMillis: number;
            minEvictableIdleTimeMillis: number;
            softMinEvictableIdleTimeMillis: number;
            numTestsPerEvictionRun: number;
            evictorShutdownTimeoutMillis: number;
            testOnCreate: boolean;
            testOnBorrow: boolean;
            testOnReturn: boolean;
            testWhileIdle: boolean;
            timeBetweenEvictionRunsMillis: number;
            evictionPolicyClassName?: string;
            blockWhenExhausted: boolean;
            jmxEnabled: boolean;
            jmxNameBase?: string;
            jmxNamePrefix?: string;
        };
    };
}
