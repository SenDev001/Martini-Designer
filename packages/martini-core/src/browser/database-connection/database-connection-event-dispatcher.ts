import { Emitter, Event } from "@theia/core";
import { injectable } from "inversify";
import { DatabaseConnectionEvent } from "../../common/database-connection/martini-database-connection-manager";
import { MartiniEndpointManagerClient } from "../../common/endpoint/martini-endpoint-manager";

export const DatabaseConnectionEventDispatcher = Symbol("DatabaseConnectionEventDispatcher");

export interface DatabaseConnectionEventDispatcher {
    onDatabaseConnectionEvent: Event<DatabaseConnectionEvent>;
}

@injectable()
export class DatabaseConnectionEventDispatcherClient
    implements MartiniEndpointManagerClient, DatabaseConnectionEventDispatcher {
    private readonly onDatabaseConnectionEventEmitter = new Emitter<DatabaseConnectionEvent>();
    readonly onDatabaseConnectionEvent = this.onDatabaseConnectionEventEmitter.event;

    onEvent(event: DatabaseConnectionEvent): void {
        this.onDatabaseConnectionEventEmitter.fire(event);
    }
}
