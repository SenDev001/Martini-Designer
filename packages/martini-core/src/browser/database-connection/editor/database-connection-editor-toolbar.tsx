import * as React from "react";
import styled from "styled-components";
import { DatabaseType, getTypeLabel } from "../../../common/database-connection/martini-database-connection-manager";
import { ToolBar, ToolBarItem } from "../../components/toolbar";
import { errorsToString, FormError } from "../../form/form";
import { CommandRegistry } from "@phosphor/commands";
import { Icon, OverlayIcon } from "../../components/icon-overlay";
import messages from "martini-messages/lib/messages";

interface DatabaseConnectionEditorToolbarProps {
    onStart?: () => void;
    onStop?: () => void;
    onTest?: () => void;
    onSearch?: () => void;
    onRevert?: () => void;
    connectionName: string;
    databaseType: DatabaseType;
    errors: FormError[],
    status: "STOPPED" | "STARTED";
}

const DatabaseConnectionEditorToolbarStyles = styled.div`
    display: grid;
    grid-template-columns: max-content 1fr max-content;
    padding: 14px;

    .title-wrapper {
        text-overflow: ellipsis;
        overflow: hidden;
        display: grid;
    }

    .title {
        padding-left: var(--theia-ui-padding);
        font-size: large;
        white-space: pre;
        min-width: 0;
        text-overflow: ellipsis;
        overflow: hidden;
        align-self: center;
    }

    .error-label {
        font-size: unset;
    }
`;

export const DatabaseConnectionEditorToolbar: React.FC<DatabaseConnectionEditorToolbarProps> = ({
    onStart,
    onStop,
    onTest,
    onSearch,
    onRevert,
    connectionName,
    databaseType,
    status,
    errors
}) => {
    let title = <div className="title-wrapper">
        <div className="title">{getTypeLabel(databaseType)} Connection - {connectionName}</div>
    </div>;
    if (errors.length !== 0) {
        const tooltip = errorsToString(errors);
        title = <div className="title-wrapper">
            <div
                className="title error-label"
                title={tooltip}
            >
                {errors[0].label}: {errors[0].message}
            </div>
        </div>;
    }

    return <DatabaseConnectionEditorToolbarStyles>
        <Icon iconClass={errors.length === 0 ? `martini-${databaseType}-icon` : "martini-error-icon"}>
            <OverlayIcon
                iconClass={status === "STOPPED" ? "martini-stopped-overlay-icon" : "martini-started-overlay-icon"}
                position="top-right"
                width={7}
            />
        </Icon>
        {title}
        <ToolBar layout="horizontal">
            <button
                className="theia-button"
                onClick={onTest}
                disabled={errors.length !== 0}
            >
                {messages.test_configuration}
            </button>
            <button className="theia-button" onClick={onRevert}>{messages.revert}</button>
            <ToolBarItem
                iconClass="martini-search-icon"
                tooltip={`Find (${CommandRegistry.formatKeystroke("Accel F")})`}
                onClick={onSearch}
            />
            <ToolBarItem
                iconClass="martini-start-icon"
                onClick={onStart}
                disabledIconClass="martini-start-icon disabled-icon"
                enabled={status === "STOPPED" && !!onStart}
            />
            <ToolBarItem
                iconClass="martini-stop-icon"
                onClick={onStop}
                disabledIconClass="martini-stop-icon disabled-icon"
                enabled={status === "STARTED" && !!onStop}
            />
        </ToolBar>
    </DatabaseConnectionEditorToolbarStyles>;
};
