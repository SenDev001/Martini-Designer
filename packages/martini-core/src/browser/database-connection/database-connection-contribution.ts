import {
    Command,
    CommandContribution,
    CommandRegistry,
    MenuContribution,
    MenuModelRegistry,
    SelectionService
} from "@theia/core";
import { ApplicationShell, CommonCommands } from "@theia/core/lib/browser";
import { inject, injectable } from "inversify";
import messages from "martini-messages/lib/messages";
import {
    DatabaseConnection,
    MartiniDatabaseConnectionManager
} from "../../common/database-connection/martini-database-connection-manager";
import { ConfirmDialog, createListMessage } from "../dialogs/dialogs";
import { DeleteCommand } from "../martini-ide-contribution";
import { NAVIGATOR_VIEW_CONTEXT_MENU } from "../navigator/martini-navigator-view-contribution";
import { ProgressService } from "../progress/progress-service";
import { DatabaseConnectionEditor } from "./editor/database-connection-editor";

export const DATABASE_CATEGORY = "Database";

export const StartDatabaseConnectionCommand: Command = {
    id: "martini.startDatabaseConnection",
    label: messages.start,
    iconClass: "martini-icon martini-database-connection-start-icon",
    category: DATABASE_CATEGORY
};

export const StopDatabaseConnectionCommand: Command = {
    id: "martini.stopDatabaseConnection",
    label: messages.stop,
    iconClass: "martini-icon martini-database-connection-stop-icon",
    category: DATABASE_CATEGORY
};

@injectable()
export class DatabaseConnectionCommandContribution implements CommandContribution {
    @inject(SelectionService)
    private readonly selectionService: SelectionService;
    @inject(ProgressService)
    private readonly progressService: ProgressService;
    @inject(MartiniDatabaseConnectionManager)
    private readonly dbConnectionManager: MartiniDatabaseConnectionManager;
    @inject(ApplicationShell)
    private readonly shell: ApplicationShell;

    registerCommands(commands: CommandRegistry): void {
        commands.registerCommand(StartDatabaseConnectionCommand, {
            execute: async () => {
                const selection = this.selectionService.selection;
                if (isDbConnectionSelection(selection)) {
                    const connections = selection.filter(c => c.status !== "STARTED");

                    const dlg = new ConfirmDialog({
                        title:
                            connections.length > 1
                                ? messages.start_multi_db_connections
                                : messages.start_single_db_connection,
                        msg:
                            connections.length <= 1
                                ? messages.start_single_db_connection_question(connections[0].name)
                                : createListMessage(
                                      messages.start_multi_db_connections_question,
                                      connections.map(c => c.name)
                                  ),
                        ok: messages.start
                    });

                    const confirmed = await dlg.open();
                    if (confirmed) {
                        await this.progressService.showProgress(messages.starting_db_connection, async progress => {
                            let count = 1;
                            for (const connection of connections) {
                                if (progress.isCancelled()) break;
                                progress.report({
                                    message: connection.name,
                                    work: { done: count++, total: connections.length }
                                });
                                await this.dbConnectionManager.start(connection.name);
                            }
                        });
                    }
                }
            },
            isEnabled: () => {
                const selection = this.selectionService.selection;
                return isDbConnectionSelection(selection) && selection.some(c => c.status !== "STARTED");
            },
            isVisible: () => {
                const selection = this.selectionService.selection;
                return isDbConnectionSelection(selection) && selection.some(c => c.status !== "STARTED");
            }
        });
        commands.registerCommand(StopDatabaseConnectionCommand, {
            execute: async () => {
                const selection = this.selectionService.selection;
                if (isDbConnectionSelection(selection)) {
                    const connections = selection.filter(c => c.status !== "STOPPED");
                    const warning = !connections.some(c => CORE_DB_CONNECTIONS.includes(c.name))
                        ? ""
                        : messages.deleting_core_db_connection_warning;
                    const dlg = new ConfirmDialog({
                        title:
                            connections.length > 1
                                ? messages.stop_multi_db_connections
                                : messages.stop_single_db_connection,
                        msg:
                            connections.length <= 1
                                ? messages.stop_single_db_connection_question(connections[0].name) + " " + warning
                                : createListMessage(
                                      messages.stop_multi_db_connections_question + " " + warning,
                                      connections.map(c => c.name)
                                  ),
                        ok: messages.stop
                    });

                    const confirmed = await dlg.open();
                    if (confirmed) {
                        await this.progressService.showProgress(messages.stopping_db_connection, async progress => {
                            let count = 1;
                            for (const connection of connections) {
                                if (progress.isCancelled()) break;
                                progress.report({
                                    message: connection.name,
                                    work: { done: count++, total: connections.length }
                                });
                                await this.dbConnectionManager.stop(connection.name);
                            }
                        });
                    }
                }
            },
            isEnabled: () => {
                const selection = this.selectionService.selection;
                return isDbConnectionSelection(selection) && selection.some(c => c.status !== "STOPPED");
            },
            isVisible: () => {
                const selection = this.selectionService.selection;
                return isDbConnectionSelection(selection) && selection.some(c => c.status !== "STOPPED");
            }
        });
        commands.registerHandler(DeleteCommand.id, {
            execute: async () => {
                const selection = this.selectionService.selection;
                if (isDbConnectionSelection(selection)) {
                    const connections = selection.filter(c => !CORE_DB_CONNECTIONS.includes(c.name));

                    const dlg = new ConfirmDialog({
                        title:
                            connections.length > 1
                                ? messages.delete_multi_db_connections
                                : messages.delete_single_db_connection,
                        msg:
                            connections.length <= 1
                                ? messages.delete_single_db_connection_question(connections[0].name)
                                : createListMessage(
                                      messages.delete_multi_db_connections_question,
                                      connections.map(c => c.name)
                                  ),
                        ok: messages.stop
                    });

                    const confirmed = await dlg.open();
                    if (confirmed) {
                        await this.progressService.showProgress(messages.deleting_db_connection, async progress => {
                            let count = 1;
                            for (const connection of connections) {
                                if (progress.isCancelled()) break;
                                progress.report({
                                    message: connection.name,
                                    work: { done: count++, total: connections.length }
                                });
                                await this.dbConnectionManager.delete(connection.name);
                            }
                        });
                    }
                }
            },
            isEnabled: () => {
                const selection = this.selectionService.selection;
                return isDbConnectionSelection(selection) && selection.some(c => !CORE_DB_CONNECTIONS.includes(c.name));
            },
            isVisible: () => {
                const selection = this.selectionService.selection;
                return isDbConnectionSelection(selection) && selection.some(c => !CORE_DB_CONNECTIONS.includes(c.name));
            }
        });
        const findHandler = {
            execute: () => {
                if (this.shell.currentWidget instanceof DatabaseConnectionEditor)
                    this.shell.currentWidget.showSearchBox();
            },
            isEnabled: () => this.shell.currentWidget instanceof DatabaseConnectionEditor
        };
        commands.registerHandler(CommonCommands.FIND.id, findHandler);
        commands.registerHandler("actions.find", findHandler);
    }
}

export const isDbConnectionSelection = (selection: any): selection is DatabaseConnection[] => {
    return selection instanceof Array && selection.find(e => DatabaseConnection.is(e));
};

const CORE_DB_CONNECTIONS = ["coder", "flux", "tracker", "config", "monitor"];

@injectable()
export class DatabaseConnectionMenuContribution implements MenuContribution {
    registerMenus(menus: MenuModelRegistry): void {
        menus.registerMenuAction(NAVIGATOR_VIEW_CONTEXT_MENU, {
            commandId: StartDatabaseConnectionCommand.id
        });
        menus.registerMenuAction(NAVIGATOR_VIEW_CONTEXT_MENU, {
            commandId: StopDatabaseConnectionCommand.id
        });
    }
}
