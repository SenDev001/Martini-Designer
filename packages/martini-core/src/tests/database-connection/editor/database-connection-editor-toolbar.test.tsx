import * as React from "react";
import * as Enzyme from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import * as Renderer from "react-test-renderer";
import { DatabaseConnectionEditorToolbar } from "../../../browser/database-connection/editor/database-connection-editor-toolbar";
import { DatabaseType } from "../../../common/database-connection/martini-database-connection-manager";
import { noop } from "lodash";

Enzyme.configure({ adapter: new Adapter() });

test("DatabaseConnectionEditorToolbar should be rendered", () => {
    const component = Renderer.create(<DatabaseConnectionEditorToolbar
        connectionName="test"
        onStart={noop}
        onStop={noop}
        databaseType={DatabaseType.JDBC}
        errors={[]}
        status="STARTED"
    />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("DatabaseConnectionEditorToolbar should be rendered with start and stop disabled", () => {
    const component = Renderer.create(<DatabaseConnectionEditorToolbar
        connectionName="test"
        databaseType={DatabaseType.JDBC}
        errors={[]}
        status="STARTED"
    />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("DatabaseConnectionEditorToolbar should be rendered with errors", () => {
    const component = Renderer.create(<DatabaseConnectionEditorToolbar
        connectionName="test"
        onStart={noop}
        onStop={noop}
        databaseType={DatabaseType.JDBC}
        errors={[{ label: "URL", message: "Cannot be blank." }]}
        status="STARTED"
    />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
