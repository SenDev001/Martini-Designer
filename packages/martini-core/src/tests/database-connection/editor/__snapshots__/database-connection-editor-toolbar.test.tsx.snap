// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatabaseConnectionEditorToolbar should be rendered 1`] = `
<div
  className="sc-AxirZ hEJXgT"
>
  <div
    className="martini-tree-icon martini-jdbc-icon"
    style={
      Object {
        "marginRight": "4px",
        "position": "relative",
      }
    }
  >
    <div
      className="martini-started-overlay-icon"
      style={
        Object {
          "backgroundPosition": "center center",
          "backgroundRepeat": "no-repeat",
          "backgroundSize": "7px 8px",
          "height": 8,
          "position": "absolute",
          "right": "0",
          "top": "0",
          "width": 7,
        }
      }
    />
  </div>
  <div
    className="title-wrapper"
  >
    <div
      className="title"
    >
      JDBC
       Connection - 
      test
    </div>
  </div>
  <div
    className="sc-AxjAm Zaixe"
    style={
      Object {
        "gridTemplateColumns": "repeat(5, max-content)",
        "height": undefined,
      }
    }
  >
    <button
      className="theia-button"
      disabled={false}
    >
      Test Configuration
    </button>
    <button
      className="theia-button"
    >
      Revert
    </button>
    <div
      className="item item-enabled"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseOut={[Function]}
      onMouseUp={[Function]}
      title="Find (Ctrl+F)"
    >
      <div
        className="martini-icon martini-search-icon"
        style={
          Object {
            "width": "16px",
          }
        }
      />
    </div>
    <div
      className="item"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseOut={[Function]}
      onMouseUp={[Function]}
    >
      <div
        className="martini-icon martini-start-icon disabled-icon"
        style={
          Object {
            "width": "16px",
          }
        }
      />
    </div>
    <div
      className="item item-enabled"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseOut={[Function]}
      onMouseUp={[Function]}
    >
      <div
        className="martini-icon martini-stop-icon"
        style={
          Object {
            "width": "16px",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`DatabaseConnectionEditorToolbar should be rendered with errors 1`] = `
<div
  className="sc-AxirZ hEJXgT"
>
  <div
    className="martini-tree-icon martini-error-icon"
    style={
      Object {
        "marginRight": "4px",
        "position": "relative",
      }
    }
  >
    <div
      className="martini-started-overlay-icon"
      style={
        Object {
          "backgroundPosition": "center center",
          "backgroundRepeat": "no-repeat",
          "backgroundSize": "7px 8px",
          "height": 8,
          "position": "absolute",
          "right": "0",
          "top": "0",
          "width": 7,
        }
      }
    />
  </div>
  <div
    className="title-wrapper"
  >
    <div
      className="title error-label"
      title="URL: Cannot be blank."
    >
      URL
      : 
      Cannot be blank.
    </div>
  </div>
  <div
    className="sc-AxjAm Zaixe"
    style={
      Object {
        "gridTemplateColumns": "repeat(5, max-content)",
        "height": undefined,
      }
    }
  >
    <button
      className="theia-button"
      disabled={true}
    >
      Test Configuration
    </button>
    <button
      className="theia-button"
    >
      Revert
    </button>
    <div
      className="item item-enabled"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseOut={[Function]}
      onMouseUp={[Function]}
      title="Find (Ctrl+F)"
    >
      <div
        className="martini-icon martini-search-icon"
        style={
          Object {
            "width": "16px",
          }
        }
      />
    </div>
    <div
      className="item"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseOut={[Function]}
      onMouseUp={[Function]}
    >
      <div
        className="martini-icon martini-start-icon disabled-icon"
        style={
          Object {
            "width": "16px",
          }
        }
      />
    </div>
    <div
      className="item item-enabled"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseOut={[Function]}
      onMouseUp={[Function]}
    >
      <div
        className="martini-icon martini-stop-icon"
        style={
          Object {
            "width": "16px",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`DatabaseConnectionEditorToolbar should be rendered with start and stop disabled 1`] = `
<div
  className="sc-AxirZ hEJXgT"
>
  <div
    className="martini-tree-icon martini-jdbc-icon"
    style={
      Object {
        "marginRight": "4px",
        "position": "relative",
      }
    }
  >
    <div
      className="martini-started-overlay-icon"
      style={
        Object {
          "backgroundPosition": "center center",
          "backgroundRepeat": "no-repeat",
          "backgroundSize": "7px 8px",
          "height": 8,
          "position": "absolute",
          "right": "0",
          "top": "0",
          "width": 7,
        }
      }
    />
  </div>
  <div
    className="title-wrapper"
  >
    <div
      className="title"
    >
      JDBC
       Connection - 
      test
    </div>
  </div>
  <div
    className="sc-AxjAm Zaixe"
    style={
      Object {
        "gridTemplateColumns": "repeat(5, max-content)",
        "height": undefined,
      }
    }
  >
    <button
      className="theia-button"
      disabled={false}
    >
      Test Configuration
    </button>
    <button
      className="theia-button"
    >
      Revert
    </button>
    <div
      className="item item-enabled"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseOut={[Function]}
      onMouseUp={[Function]}
      title="Find (Ctrl+F)"
    >
      <div
        className="martini-icon martini-search-icon"
        style={
          Object {
            "width": "16px",
          }
        }
      />
    </div>
    <div
      className="item"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseOut={[Function]}
      onMouseUp={[Function]}
    >
      <div
        className="martini-icon martini-start-icon disabled-icon"
        style={
          Object {
            "width": "16px",
          }
        }
      />
    </div>
    <div
      className="item"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseOut={[Function]}
      onMouseUp={[Function]}
    >
      <div
        className="martini-icon martini-stop-icon disabled-icon"
        style={
          Object {
            "width": "16px",
          }
        }
      />
    </div>
  </div>
</div>
`;
